source/fuzz/fuzzer_pass_wrap_vector_synonym.cpp

```diff
@@ -0,0 +1,164 @@
+        // vector.
+        uint32_t position = std::rand() % component_count;
+
+        // target ids are the two scalar ids from the original instruction.
+        uint32_t target_id1 = instruction_iterator->GetSingleWordOperand(1);
+        uint32_t target_id2 = instruction_iterator->GetSingleWordOperand(2);
+
+        // Stores the ids of scalar constants.
+        std::vector<uint32_t> vec1_components;
+        std::vector<uint32_t> vec2_components;
+
+        // Width is specified in the first index for either OpTypeInt or
+        // OpTypeFloat.
+        uint32_t width = GetIRContext()
+                             ->get_def_use_mgr()
+                             ->GetDef(scalar_type_id)
+                             ->GetSingleWordOperand(0);
+        // Get the scalar type.
+        auto type = GetIRContext()->get_type_mgr()->GetType(scalar_type_id);
+
+        // Whether the constant is signed, not used for float type so default
+        // set to true.
+        bool is_signed =
+            type->AsInteger() ? type->AsInteger()->IsSigned() : true;
+
+        // Populate components based on vector type and size.
+        for (uint32_t i = 0; i < component_count; ++i) {
+          if (i == position) {
+            vec1_components.emplace_back(target_id1);
+            vec2_components.emplace_back(target_id2);
+          } else {
+            switch (instruction_iterator->opcode()) {
+              case SpvOpTypeInt: {
+                AddRandomIntConstant(vec1_components, width, is_signed);
+                AddRandomIntConstant(vec2_components, width, is_signed);
+                break;
+              }
+              case SpvOpTypeFloat: {
+                AddRandomFloatConstant(vec1_components, width);
+                AddRandomFloatConstant(vec2_components, width);
+                break;
+              }
+              default:
+                assert(false &&
+                       "Instruction opcode must be a valid numeric constant "
+                       "type.");
+            }
+          }
+        }
+        // Add two OpCompositeConstruct to the module with result id returned.
+        uint32_t result_id1 = AddNewVecNType(scalar_type_id, vec1_components,
+                                             instruction_descriptor);
+        uint32_t result_id2 = AddNewVecNType(scalar_type_id, vec2_components,
+                                             instruction_descriptor);
+
+        // Apply transformation to do vector operation and add synonym between
+        // the result vector id and the id of the original instruction.
+        ApplyTransformation(TransformationWrapVectorSynonym(
+            instruction_iterator->result_id(), result_id1, result_id2,
+            GetFuzzerContext()->GetFreshId(), vec_type_id, position));
+      });
+}
+
+uint32_t FuzzerPassWrapVectorSynonym::AddNewVecNType(
+    uint32_t composite_type_id, std::vector<uint32_t> component,
```

Comment:
Please use `const std::vector<uint32_t>&`. 

In general, however, it might be even better to remove this function and call `ApplyTransformation(TransformationCompositeConstruct(...))` in the `Apply` method directly. This is not that much duplication and will probably help to improve code clarity :) Up to you though.

Reply:
If you decide to keep it, it will be better to come up with a better name for it (e.g. `ConstructVectorComposite`).

Reply:
Done.